Add UI for updating subgroups of a group.

sahil839 opened the pull request:
<!-- Describe your pull request here.-->
- First two commits are for refactoring.
- Third commit is to fix the behavior of pressing enter inside the pill input.
- Fourth commit is to update code for checking whether the user being added to a group is already a member to consider only direct members.
- Fifth commit is to add UI for updating subgroups of existing groups.
- Sixth commit is to add UI for adding subgroups while creating groups.

<!-- Issue link, or clear description.-->

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**
![create-with-subgroups](https://github.com/user-attachments/assets/1fea57ae-bb35-4764-8ce5-5f2e261b9937)
![subroups-ui](https://github.com/user-attachments/assets/1dbbc730-8b7f-445e-96fe-d3c4a4cdb506)

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

sahil839 commented on web/src/user_group_edit.js:
```diff
@@ -285,6 +285,18 @@ function update_group_membership_button(group_id) {
+    // update members list if currently rendered.
+    if (is_editing_group(group_id)) {
+        user_group_edit_members.update_member_list_widget(group);
+    }
```
There is still some work needed here for live-update like permissions to manage a group can also change if a subgroup is added to a group.

timabbott replied:
Do you mean specifically that whether you have permission to edit the other settings may need to live-update? Yeah, that makes sense. Hopefully that code does not end up super fiddly.

sahil839 replied:
Yes. Adding or removing a subgroup may result in changing permissions if the modified group is used for a setting.

timabbott replied:
Hmm, yeah. That may prove to be messy to trace through all the things that might need to rerender. 
Let's open an issue to track that, as it may be a longer project that needs some strategy.

timabbott replied:
Opened #32089 to track this.

sahil839 commented on web/src/user_group_edit_members.ts:
```diff
@@ -66,17 +75,82 @@ function format_member_list_elem(person: User): string {
+}
+
+function sort_group_member_email(a: User | UserGroup, b: User | UserGroup): number {
+    if ("user_id" in a && "user_id" in b) {
```
Inintially had these conditions as `a.user_id !== undefined && b.user_id !== undefined` but linter gave an error.

timabbott replied:
This is a little more precise than the original; ultimately the check is whether this is a User or UserGroup.

sahil839 commented on web/src/user_group_edit_members.ts:
```diff
@@ -258,17 +352,56 @@ function add_new_members({pill_user_ids}: {pill_user_ids: number[]}): void {
+        const ignored_already_added_subgroup_ids = [...already_added_subgroups];
+        ignored_already_added_subgroups = ignored_already_added_subgroup_ids.map((group_id) =>
+            user_groups.get_user_group_from_id(group_id),
+        );
```
This can probably be deduplicated. Or maybe we can split some code to different functions for readability. Will look into this.

timabbott replied:
Yeah this could be nicer but it's basically fine from a readability standpoint.

sahil839 commented:
There is still some work left -
- Some live-update work like enabling/disabling the settings if permission changes due to updating subgroups.
- Currently the typeahead shows all the user groups. We can hide the groups which we do not allow to follow the DAG constraint.

Can also de-duplicate some parts of code. Might still be good for a manual testing by @alya.

alya commented:
I see this when trying to add a group to itself:

![Screenshot 2024-10-18 at 10 09 34](https://github.com/user-attachments/assets/84ddd267-443c-4f09-9b9a-6ec795d80c35)

Will taking it out of the typeahead would be covered by the DAG constraint work? Or maybe there's a separate change to removing the group itself + subgroups that already exist?

We should also make the error message more user-friendly. I'm not sure exactly where the error is used, so I dunno if @timabbott would find it easier to suggest a string.

alya commented:
I actually go this when trying to add a group for the second time:

![Screenshot 2024-10-18 at 10 15 31](https://github.com/user-attachments/assets/5dea25bd-7533-4e7d-bb87-6896e3c431e5)

alya commented:
#31850 also notes a change to the placeholder text.

timabbott commented on web/src/user_group_edit_members.ts:
```diff
@@ -373,8 +509,57 @@ function remove_member({
+    const group = user_groups.get_user_group_from_id(current_group_id);
+    if (!group) {
+        return;
+    }
```
What does this case represent on the UI side?

sahil839 replied:
I thought this is present for avoiding the typescript errors in case when `get_user_group_from_id` returns undefined for invalid IDs but that throws error for such cases, so can remove it.

timabbott commented on web/src/user_group_edit_members.ts:
```diff
@@ -373,8 +509,57 @@ function remove_member({
+        });
+    }
+
+    do_remove_subgroup_from_group();
```
Why do we define the `do_remove_subgroup_from_group` if it's just going to be called once?

timabbott commented on web/styles/input_pill.css:
```diff
@@ -284,6 +284,10 @@
 }
 
+.display_only_group_pill .zulip-icon-triple-users {
+    font-size: 19px;
```
Should this be in `em` to scale properly with font size?

sahil839 replied:
I tried using em value of `1.19em` (computed from using 19px for 16px font size), but feels like 19px is better visually -
| | Using 19px | Using 1.19em |
| ----------- | --------------- | ------------------ |
| For 14/122px | ![compact-19px](https://github.com/user-attachments/assets/ffd4f699-d813-4039-98b9-cf48cce5bc9a) | ![compact-em](https://github.com/user-attachments/assets/4f4542ef-481f-45b8-a542-6876da536559) |
| For 16/140px | ![19px](https://github.com/user-attachments/assets/f27b2762-f801-4b30-8b50-923ea15750d6) | ![em](https://github.com/user-attachments/assets/e1f897c7-696b-488a-8022-5ddca2ed1f04) |

sahil839 commented:
> Will taking it out of the typeahead would be covered by the DAG constraint work? Or maybe there's a separate change to removing the group itself + subgroups that already exist?

Yes that will be covered.

sahil839 commented:
> We should also make the error message more user-friendly. I'm not sure exactly where the error is used, so I dunno if @timabbott would find it easier to suggest a string.

It is essentially showing that this will break the DAG constraint. But may be we can have different error messages for cases like trying to add the group itself as subgroup. Maybe we can just have these different error messages in client only, i.e. we check this at the client itself.

sahil839 commented:
- Fixed the bug when trying to add a group second time. It now shows a message below the input pill container.
- Also, fixed the issues with group popovers which I forgot to do before.
- Updated the placeholder as well.

timabbott commented:
> It is essentially showing that this will break the DAG constraint. But may be we can have different error messages for cases like trying to add the group itself as subgroup. Maybe we can just have these different error messages in client only, i.e. we check this at the client itself.

Yeah, let's just open a follow-up issue to track improving the user experience for trying to violate the DAG constraint. I don't think it's a thing users will run into often... and probably the client-side thing we will end up doing might be to just not offer the current group's supergroups as options in the pill widget, rather than anything with strings.

timabbott commented:
(Leaving this for tomorrow, since I merged a PR from Shubham conflicting with this today.)

sahil839 commented:
Rebased the PR.

Also, added a commit to fix the buggy behavior when pressing enter inside the pill container.

timabbott commented:
This looks great, merged, thanks @sahil839! I opened a couple follow-ups:
- #32087 to track not offering supergroups of the current group in the members picker. This seems like a high priority because it's probably easy.
- #32089 to track the live-update of permissions issue noted above. While it is valuable to fix, I think it's lower priority than getting the rest of the system fully working.
